proto/src/rr/dnssec/nsec3.rs

```diff
@@ -287,13 +288,13 @@ fn test_known_hashes() {
 #[cfg(test)]
 #[cfg(any(feature = "openssl", feature = "ring"))]
 fn hash_with_base32(name: &str) -> String {
-    use data_encoding::base32hex;
+    use data_encoding::BASE32HEX;
 
     // NSEC3PARAM 1 0 12 aabbccdd
     let known_name = Name::parse(name, Some(&Name::new())).unwrap();
     let known_salt = [0xAAu8, 0xBBu8, 0xCCu8, 0xDDu8];
     let hash = Nsec3HashAlgorithm::SHA1
         .hash(&known_salt, &known_name, 12)
         .unwrap();
-    base32hex::encode(hash.as_ref()).to_lowercase()
+    BASE32HEX.encode(hash.as_ref()).to_lowercase()
```

Comment:
I am currently adding support for DNSSEC base32 in data-encoding [1]. But before doing so I want to make sure that I understand RFC5155 [2] correctly. Is my reading correct that DNSSEC base32 must satisfy the following properties?
- Extended hex alphabet "0123456789abcdefghijklmnopqrstuv".
- Case-insensitive decoding.
- No padding.

Note that it is already possible to use this encoding without waiting for at least `data-encoding:2.1.0`. However it requires to depend (here it would be a dev-dependency) on `lazy_static` and the encoding would be created at run-time (at first use) instead of compile-time.

[1] https://github.com/ia0/data-encoding/pull/11
[2] https://tools.ietf.org/html/rfc5155

Reply:
> I am currently adding support for DNSSEC base32 in data-encoding [1]

great!

>  Is my reading correct that DNSSEC base32 must satisfy the following properties?

Yes, I believe your reading of the alphabet is current, rfc5155 references this rfc directly: https://tools.ietf.org/html/rfc4648#section-7. This note is in rfc5155: `Note that trailing padding characters
      ("=") are not used in the NSEC3 specification.`

It also references that the canonical order and case insensitivity from rfc4034 is to be used: https://tools.ietf.org/html/rfc4034#section-6.1

So you're understanding is correct.

> However it requires to depend (here it would be a dev-dependency) on lazy_static and the encoding would be created at run-time (at first use) instead of compile-time.

I'm not 100% clear on what you're stipulating here. Is there a reason the alphabet can't be used at compile time only?

Reply:
> So you're understanding is correct.

Ok thanks! I'll add support for `BASE32_DNSSEC` then.

> I'm not 100% clear on what you're stipulating here. Is there a reason the alphabet can't be used at compile time only?

I am referring to using something like this:
```
lazy_static! {
    static ref BASE32_DNSSEC: Encoding = {
        let mut spec = Specification::new();
        // Extended hex alphabet.
        spec.symbols.push_str("0123456789abcdefghijklmnopqrstuv");
        // Case-insensitive.
        spec.translate.from.push_str("ABCDEFGHIJKLMNOPQRSTUV");
        spec.translate.to.push_str("abcdefghijklmnopqrstuv");
        // Unpadded.
        assert_eq!(spec.padding, None);
        spec.encoding().unwrap()
    };
}
```
I currently don't provide a way to build an encoding at compile-time because I couldn't find a simple and safe way to do so. I considered the following options:
- Using `const fn`: this would be the ideal solution but they are feature-gated and not expressive enough.
- Making the `Encoding` implementation public (not great) and:
    - Asking the users to generate the encoding content manually (like I currently do internally [1]) or through a `build.rs`.
    - Providing a procedural macro to generate the content.

However, this would be necessary only until `BASE32_DNSSEC` is released (probably in `2.1.0`), since the constant encodings provided by `data-encoding` are pre-compiled (a test guarantees that the pre-compiled version is up-to-date [2]).

Since the current usage is only in tests, the performance and allocation costs of `str::to_lowercase` do probably not matter. But if they do, I can use `lazy_static` until `data-encoding:2.1.0` is released.

[1] https://github.com/ia0/data-encoding/blob/6efb67698f424012507f09e508fc5305ddb98459/lib/src/lib.rs#L2104-L2136
[2] https://github.com/ia0/data-encoding/blob/6efb67698f424012507f09e508fc5305ddb98459/lib/src/lib.rs#L2088-L2097